### PR TITLE
Full featured property based Animation System

### PR DESCRIPTION
**Pull request by lassade**
Any type that implements `bevy_reflect::Struct` can have their attributes animated.

Has a safety restriction only attributes that are within the struct memory limits can be addressed and animated,
this means you can't animate items behind a `Vec`, array types could be animated but nested properties implementation
is missing.

The system main goal is to provide an low level api to play and blend animations.
That been said an animation state machine should be built by the user or provided by a external crate.

## Features

1. Property based animation

```rust
let mut clip = Clip::default();
/// A property is defined by the entity path followed by the component property
clip.add_curve_at_path("/Node1/Node2@Transform.translation", track_data);
```

2. Animate any type using bevy_reflect

```rust
#[derive(Debug, Reflect)]
#[reflect(Component)]
struct CustomComponent {
    value: f32,
    point: Vec3,
    #[reflect(ignore)]
    non_animated: Vec3,
}

// During app startup
app_builder.register_animated_component::<bevy_transform::prelude::Transform>();

let mut clip = Clip::default();
clip.add_curve_at_path("@CustomComponent.value", ...);
clip.add_curve_at_path("@CustomComponent.point", ...);
```

4. Animate assets by swapping handles or animate their properties

```rust
// During app startup
app_builder.register_animated_asset::<bevy_pbr::prelude::StandardMaterial>();

let mut clip = Clip::default();
// Swap material with
clip.add_curve_at_path("@Handle<StandardMaterial>", ...);
// Change the current material properties
clip.add_curve_at_path("@Handle<StandardMaterial>.albedo", ...);
clip.add_curve_at_path("@Handle<StandardMaterial>.albedo_texture", ...);
```

5. GLTF support

You can load any animation clips embedded within a GLTF file

```rust
// Loads the skinned mesh ready to be animated with all of its clips (if any)
let char_scene_handle = asset_server.load("models/character_medium/character_medium.gltf");
// Returns a single clip from the GLTF
let clip_handle = asset_server.load("models/character_medium/idle.gltf#Anim0");
```

6. Animation Blending

It's possible to blend `bool`, `f32`, `Vec2`, `Vec3`, `Vec4`, `Quat`, `Handle<T>`, `Option<T>` when `T` can also be blended, and more.

7. Animation clips defined by a named hierarchy

Each `Animator` can only animate children entities, this is similar to other engines like Unity;

Each clip have it's own hierarchy space, this allows clips to be reused for different setups that
share some similarities. 

Note that this isn't a humanoid retargeting system, but it could give you some good results is some particular situations;

8. Mostly safe :p

This crate uses a couple `unsafe`'s blocks, the safety rules of these blocks should be guaranteed
at within the scope of a function or file (in the worst case) this way changes can be more easily
made without the fear braking random safety conditions, as a bonus they should be asserted
by code whenever possible;

**The unsafe code still needs a review** 

9. Fast TM

a) Each component animation runs in parallel from each other
b) Keyframe cursor, allows for faster sample rates by remembering the last keyframe it was last frame
c) NLerp only and you wont notice the difference, hopefully ...

10. Skinned Mesh support out of the box

Simple and naive implementation, replaces the vertex shader from the `bevy_pbr::prelude::StandardMaterial`.
Further work should done along side with the rendering, but hey it works!  

11. Additive Blending

Supports additive blending as defined by the ACL mode Additive 1, that can be found
[here](https://github.com/nfrechette/acl/blob/develop/docs/additive_clips.md)

12. SIMD support (REMOVED for the time been)

Clips can now be packed, a packed clip will contain tracks that are sampled using wide types from the `ultraviolet` crate
thus resulting in a greater sampling throughput.

## Missing Features

1. Nested properties

Right now only the top most properties of each struct can be animated,
this means `"@CustomComponent.point.x"` isn't valid;

2. Masks

3. Animation events (simple and interval)

4. Morph Targets

5. Dedicated `Transform` animator system

`Transform` is the most common animated component in any game, its the only animated component registered
by default and it should receive a custom animator system to squeeze every single us that is possible,
this custom system should be also be capable of performing SIMD blending;

6. Root Motion

## Terminology and Internals

A Clip is made of a collection of Tracks, each Track animates a single property in a single entity; The animator works by layering multiple Clips on top of each other in order to evaluate the final animation; So each Layer have information about how to play a single Clip, like time and weight as well some other complicated stuff to make everything work;

Each Clip  stores its own hierarchy of entities that should be animated, each each entity in this hierarchy can be associated with one or more properties; Each property is defined by a combination of path (`Transform.translation`) and type (`Vec3`)
and contains a list of all the samplers paired with their respective target entities;

A sampler evaluates the state of a property at any given time. The `Track` is the only sampler currently implemented
and comes in 3 flavors.

1. `TrackFixed` performant in any operation range, requires a keyframe for every frame of the animation duration;

2. `TrackVariableLinear`, can be used with linear keyframe reduction technique to save memory at runtime,
most performant when sampled at a higher fps than the track sample rate. 

3. `TrackVariable` very similar to `TrackVariableLinear` but beyond linear it supports step and catmull-rom as interpolations methods,
because of that it will have a lower performance and require and extra runtime memory to store the keyframes tangents;
This track is similar to `AnimationCurve` from Unity engine;

## Safety status

Assume that `bevy_reflect::Struct` was correctly implemented

For that to happen unsafe blocks are used, but their safety constrains should be
self contained within the same file or even with the same function block, with one
single exception.

Asside from that an animator system can be implemented with only safe code for any
type that you may want, its a lot of code but mostly is boiler plates. Optionally you can
add two extra unsafe blocks to improve performance (with very little safety conditions);

### Why Hierarchy and not Entity List?

I wanted that the animation system to be the sole responsible of finding and assigning any animated entities, so I opted for using a Hierarchy instead of a entity list; This of course will make the animation system for simple to use, further more here a list of pros and cons of using an entity list:

#### Pros
 - Can limit to only the entities that we really want to animate, from deep nested hierarchies
 - Entities names don't matter
 - Easer to design a LayerMask around it
 - Bit less memory

#### Cons
 - Harder to modify and manipulate clips
 - Harder to find missing entities
 - Harder to mix clips with different hierarchies
 - Hierarchy don't matter which can lead to a giant mess, and makes impossible update many animators and their system in parallel (using hierarchy will be harder but not impossible)

## Update Apr 11

To simplify the work as suggested to me (sry forgot who) I broke this into a couple of smaller PR's:
 - [ ] #1837
 - [x] #1831
 - [ ] #1830
 - [ ] #1870, right now colors can't be animated anymore
 - [ ] #2215 #2216
 
1. Removed SIMD support to keep it simple
2. I will try to push this for bevy 0.6
3. I need to do a bit of clean up, organize a bit, some huge files to break down before making this a full pr,

### CODE REVIEW

**Comment by lassade** [comment]
Hello everyone,

I been working on this animation system for a while now, but I lack the time right now to finish it
so to anyone interested here is a few things that I need help with before making this into a full PR:

1. Merge the most recent commit (any changes to `bevy_log` crate must be reverted)
2. Added more examples to how animate material/assets properties and sprite/mesh swap
3. A review of course will be great, I do use a fair share `unsafe`'s blocks that could use some extra checks (if they could removed without compromise the performance will be great)

**Comment by kirawi** [comment]
Holy sh*t.

File: crates/bevy_animation/src/interpolate/utils.rs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+/// Quake 3 fast inverse sqrt
+///
+/// Took this one from piston: https://github.com/PistonDevelopers/skeletal_animation
+#[inline(always)]
+pub fn inv_sqrt(x: f32) -> f32 {
+    let x2: f32 = x * 0.5;
+    let mut y: f32 = x;
+
+    let mut i: i32 = unsafe { std::mem::transmute(y) };
```
Comment (bjorn3) [inline crates/bevy_animation/src/interpolate/utils.rs:46]: f32::to_bits is safe

File: crates/bevy_animation/src/tracks/fixed.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            // Overflow clamp
+            return self.keyframes[f_n].clone();
+        }
+
+        // Lerp the value
+        // SAFETY: bounds checks are performed in the lines above
+        unsafe {
+            T::lerp(
+                self.keyframes.get_unchecked(f),
+                self.keyframes.get_unchecked(f + 1),
```
Comment (bjorn3) [inline crates/bevy_animation/src/tracks/fixed.rs:111]: Please use regular indexing here. If perf becomes an issue, it can always be changed. LLVM may even be able to optimize the bounds checks away.

File: crates/bevy_animation/src/tracks/n.rs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        impl ValueN for $t {
+            type Value = $o;
+
+            type Lanes = [u16; $s];
+
+            type Outputs = [Self::Value; $s];
+
+            #[inline(always)]
+            fn pack(values: Self::Outputs) -> Self {
+                unsafe { transmute::<_, [$i; $s]>(values).into() }
```
Comment (bjorn3) [inline crates/bevy_animation/src/tracks/n.rs:167]: For Vec3 this packs the padding too, which is UB.

File: crates/bevy_animation/src/tracks/n.rs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+valuen!(Vec4x8, ultraviolet::Vec4, Vec4, 8);
+
+impl ValueN for Quatx4 {
+    type Value = Quat;
+
+    type Lanes = [u16; 4];
+
+    type Outputs = [Self::Value; 4];
+
+    #[inline(always)]
```
Comment (bjorn3) [inline crates/bevy_animation/src/tracks/n.rs:196]: Please use `#[inline]` instead. That allows LLVM to not inline in debug mode or when it is consideres to not have an advantage.

File: crates/bevy_animation/src/blending.rs [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+    bits: Vec<Mask>,
+    weight_slots: usize,
+    weights: Vec<f32>,
+}
+
+impl AnimatorBlending {
+    #[inline(always)]
+    pub fn begin_blending(&mut self, entities: usize, weight_slots: usize) -> AnimatorBlendGroup {
+        self.bits.clear();
+        self.bits.resize(entities, 0);
```
Comment (Archina) [inline crates/bevy_animation/src/blending.rs:29]: nit: I think i would try to be clearer in my intensions by using `self.bits = vec![0; entities];`

Same as for weights below. I am sure you are trying to optimize and reuse memory but with changing sizes I feel like this may not be the case.

File: crates/bevy_animation/src/_custom.rs [hunk 6]
```diff
@@ -0,0 +1,2 @@
+// TODO: Custom implementations for types that are very common like Transforms!
+// TODO: Custom implementation for MorphTargets or something like it
```
Comment (Archina) [inline crates/bevy_animation/src/_custom.rs:2]: This sounds awesome but I would just remove the stumps for now.

File: crates/bevy_animation/src/_experimental.rs [hunk 7]
```diff
@@ -0,0 +1,142 @@
+//! Contains unused experimental ideas
```
Comment (Archina) [inline crates/bevy_animation/src/_experimental.rs:1]: Could you gives us a hint on how experimental your experimentals are? So far I would see them as none functional but I would have to take a closer look on a checkout.

File: crates/bevy_animation/src/_props.rs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    pub const fn owned(name: String) -> Self {
+        Prop {
+            name: Cow::Owned(name),
+            _marker: PhantomData,
+        }
+    }
+}
+
+// ? NOTE: That upsets me, it should be auto generated path
```
Comment (Archina) [inline crates/bevy_animation/src/_props.rs:28]: Is this the draft to animated nested properties? I feel like this should be possible with the `syn` crate and/or a macro.

File: crates/bevy_animation/src/animator.rs [hunk 9]
```diff
@@ -0,0 +1,1411 @@
...
+        self.untyped.downcast_ref()
+    }
+
+    #[inline(always)]
+    pub fn downcast_mut<T: 'static>(&mut self) -> Option<&mut Tracks<T>> {
+        self.untyped.downcast_mut()
+    }
+}
+
+// Pros:
```
Comment (Archina) [inline crates/bevy_animation/src/animator.rs:159]: I feel like this should be part of the PR draft. It has a lot of useful information and estimations cart should see upfront.

File: crates/bevy_animation/src/animator.rs [hunk 10]
```diff
@@ -0,0 +1,1411 @@
...
+const CACHE_SIZE: usize = 128;
+#[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
+const CACHE_SIZE: usize = 64;
+
+/// Defines the number of keyframes that fits inside a cache line;
+const KEYFRAMES_PER_CACHE: usize = CACHE_SIZE / std::mem::size_of::<u16>();
+
+///////////////////////////////////////////////////////////////////////////////
+
+pub struct Tracks<T> {
```
Comment (Archina) [inline crates/bevy_animation/src/animator.rs:54]: I think the hierarchy of tracks, clips, key frames, layers is a bit lost on me/hard to follow.

Would you mind laying that structure out a bit more clearly in the PR? Also from the PR it seems like a user would mostly just interact with clips. (Thought I feel like key frames and layers may just be as important)

**Review by Archina — COMMENTED** [review]
Sorry, I am far from being through and mostly just glanced and went over the information in a meta way. I hope I can find some more time to look a bit closer at this later.

File: crates/bevy_animation/src/_experimental.rs (hunk 7 above)
Comment (lassade) [reply crates/bevy_animation/src/_experimental.rs:1]: This is just stuff that I was experimenting with nothing relevant at this stage;

File: crates/bevy_animation/src/_props.rs (hunk 8 above)
Comment (lassade) [reply crates/bevy_animation/src/_props.rs:28]: The function of this `Prop` is to associate each property with a unique type and also avoid misspells; I did tried using proc_macro with no success, thats why I removed from now

File: crates/bevy_animation/src/animator.rs (hunk 9 above)
Comment (lassade) [reply crates/bevy_animation/src/animator.rs:159]: I wanted that the animation system to be the sole responsible of finding and assigning any animated entities, so I opted for using a Hierarchy instead of a entity list;

File: crates/bevy_animation/src/animator.rs (hunk 10 above)
Comment (lassade) [reply crates/bevy_animation/src/animator.rs:54]: Clip is made of a collection of Tracks, each Track animates a single property in a single entity; The animator works by layering multiple Clips on top of each other in order to evaluate the final animation; So each Layer have information about how to play  a single Clip, like time and weight as well some other complicated stuff to make everything work;

Made more sence?

File: crates/bevy_animation/src/interpolate/utils.rs (hunk 1 above)
Comment (lassade) [reply crates/bevy_animation/src/interpolate/utils.rs:46]: Good catch `y.to_bits() as i32` should yield the same result

Comment (lassade) [reply crates/bevy_animation/src/interpolate/utils.rs:46]: Good catch `y.to_bits() as i32` should yield the same result

File: Cargo.toml [hunk 11]
```diff
@@ -362,6 +364,14 @@ crate-type = ["cdylib"]
 name = "android"
 path = "examples/android/android.rs"
 
+[[example]]
+name = "run_animation"
+path = "examples/animations/run_animation.rs"
+
+[[example]]
+name = "bench_run_idle_animation"
+path = "examples/animations/bench_run_idle_animation.rs"
```
Comment (MinerSebas) [inline Cargo.toml:373]: The examples also need to be added to the README.md in the examples folder.

Also, the `bench_run_idle_animation` doesn't exist (anymore) in this PR.

File: crates/bevy_animation/src/animator.rs (hunk 10 above)
Comment (Archina) [reply crates/bevy_animation/src/animator.rs:54]: Yes that helped a lot =) Thank you very much.

File: crates/bevy_animation/src/animator.rs (hunk 9 above)
Comment (Archina) [reply crates/bevy_animation/src/animator.rs:159]: I mean more moving the Pros and Cons into the PR draft.md, so people can pick up quickly on what you took into consideration and what compromises you have chosen.

File: crates/bevy_animation/src/blending.rs (hunk 5 above)
Comment (lassade) [reply crates/bevy_animation/src/blending.rs:29]: `self.bits = vec![0; entities];` will always allocate a new vec, while resize wont if the sizes match

Comment (Archina) [reply crates/bevy_animation/src/blending.rs:29]: Are you not forcing a mismatch in size by using `clear()` before you resize?

**Comment by lassade** [comment]
Ops sorry

File: crates/bevy_animation/src/blending.rs (hunk 5 above)
Comment (lassade) [reply crates/bevy_animation/src/blending.rs:29]: Sorry is not about the size but the vec capacity

File: crates/bevy_animation/src/_experimental.rs (hunk 7 above)
Comment (lassade) [reply crates/bevy_animation/src/_experimental.rs:1]: File removed

File: Cargo.toml (hunk 11 above)
Comment (lassade) [reply Cargo.toml:373]: Done

File: crates/bevy_animation/src/_custom.rs (hunk 6 above)
Comment (lassade) [reply crates/bevy_animation/src/_custom.rs:2]: Is there so I don't forget to do it :P

**Comment by lassade** [comment]
I merged @Davier changes and fix a small issue with the ordering of the skinning system;

I wasnt able to create some more examples because both `Mesh` and `StandardMaterial` can't be animated yet because of some restrictions that I'm not familiar with;

**Comment by ykafia** [comment]
Hi @lassade i love your work on this, is there any help i can provide to this ?

File: examples/README.md [hunk 12]
```diff
@@ -95,6 +95,14 @@ Example | File | Description
 `wireframe` | [`3d/wireframe.rs`](./3d/wireframe.rs) | Showcases wireframe rendering
 `z_sort_debug` | [`3d/z_sort_debug.rs`](./3d/z_sort_debug.rs) | Visualizes camera Z-ordering
 
+## Animations
+
+Example | File | Description
+--- | --- | ---
+`run_animation` | [`animations/run_animation.rs`](./animations/run_animation.rs) | Demonstrates how to play and blending between a idle and run animations
+[//]: # (`mesh_swap_animation` | [`animations/mesh_swap_animation.rs`](./animations/mesh_swap_animation.rs) | Demonstrates hot to create your own animation clip that swaps meshes (or any other asset handle))
+[//]: # (`material_animation` | [`animations/material_animation.rs`](./animations/material_animation.rs) | Illustrate how to animate material properties (or any other asset property))
```
Comment (Trouv) [inline examples/README.md:104]:
```suggestion
`run_animation` | [`animations/run_animation.rs`](./animations/run_animation.rs) | Demonstrates how to play, and blend between, an idle animation and a run animation
[//]: # (`mesh_swap_animation` | [`animations/mesh_swap_animation.rs`](./animations/mesh_swap_animation.rs) | Demonstrates how to create your own animation clip that swaps meshes (or any other asset handle))
[//]: # (`material_animation` | [`animations/material_animation.rs`](./animations/material_animation.rs) | Illustrates how to animate material properties (or any other asset property))
```

**Comment by alice-i-cecile** [comment]
Closing this out; this has been incredibly valuable to lay the design foundations of our animation solutions.
